### PR TITLE
add data_source_dashboards

### PR DESCRIPTION
**Pull request by justinTM**
builds haven't been initiating so trying new fresh branch

### CODE REVIEW

**Comment by justinTM** [comment]
@inkel builds don't appear to be working

**Comment by inkel** [comment]
Hi, @justinTM, thanks again for your contributions. I've got a question: how's this PR different from #378? Both have the same title and seem to be adding the same kind of data source.

Also, I'm not so sure about this one. We already have a `grafana_dashboard` data source introduced in #359, I think we should in any case extend that one to allow for returning more than one result instead.

Let's stop development on this one temporarily while we discuss it with the rest of the maintainers.

Thanks!

**Comment by justinTM** [comment]
hey @inkel , they are virtually the same PRs; i was trying to see if i could trigger the CI build using a different branch. 


i would love to pause development honestly, but unfortunately there is a vulnerability in the way grafana provisioning (with terraform) mixes with user-created dashboards: dashboards in provisioned folders will be lost forever.  the organization i'm working with has been experiencing this issue with increasing frequency lately\. 

from a development standpoint, i'm testing cURL requests directly to the API with good (sometimes [unexpected](https://github.com/grafana/grafana/pull/44743)) results. i would just use shell commands in the project but keeping everything maintained by terraform is preferable.

`data_source_dashboard.go` could be made responsible for getting all dashboards, but currently that would require refactoring the required inputs (for id, uid, etc.). the way the aws provider seems to do it -- similar to Trott's probes datasource -- is to make them separate:
- buckets https://github.com/hashicorp/terraform-provider-aws/blob/187f1659a4fef11ac314567273b5470afe6b662f/internal/service/s3/bucket_object_data_source.go
- bucket https://github.com/hashicorp/terraform-provider-aws/blob/187f1659a4fef11ac314567273b5470afe6b662f/internal/service/s3/bucket_objects_data_source.go

if i can refactor the dashboard folder search function in the client and use it successfully, i think this PR will be ready for review.

**Comment by justinTM** [comment]
@inkel wow what a relief. I finally figured it out. There's a bug in the testing framework from Hashicorp: https://github.com/hashicorp/terraform-plugin-sdk/issues/885

**Comment by justinTM** [comment]
hey @inkel our team needs this ASAP if I could bother you for a review.

**Comment by julienduchesne** [comment]
> hey @inkel our team needs this ASAP if I could bother you for a review.

Hey @justinTM. Unfortunately, the provider is only one of our responsibilities within Grafana Labs. We actively watch PRs _with passing checks_ and we'll eventually get to yours when we get the time. Thanks for your patience!

**Comment by justinTM** [comment]
the tests are passing

**Comment by justinTM** [comment]
@julienduchesne do you know of a way to utilize a forked repo without publishing to the Hashicorp Registry?

**Comment by julienduchesne** [comment]
> @julienduchesne do you know of a way to utilize a forked repo without publishing to the Hashicorp Registry?

You can write custom versions of a provider (or completely create a new one) locally

Here's what I use from my repo:
```
go build && cp ./terraform-provider-grafana ~/.terraform.d/plugins/registry.terraform.io/grafana/grafana/999.999.999/darwin_amd64/terraform-provider-grafana
```

Then my Grafana config looks like this: 
```
terraform {
  required_providers {
    grafana = {
      source = "grafana/grafana"
    }
  }
}

provider "grafana" {
  url  = "http://localhost:3000"
  auth = "admin:admin"
}
```

**Comment by inkel** [comment]
As an alternative, if you use `terraform` CLI tool you can [override the provider installation](https://www.terraform.io/cli/config/config-file#development-overrides-for-provider-developers) via a configuration file.

File: go.mod
Comment (inkel) [inline go.mod]: Please do not commit `replace` statements in `go.mod`. Hopefully once Go 1.18 is released we will have [workspaces](https://pkg.go.dev/cmd/go@master#hdr-Workspace_maintenance) that allow having this `replace` statements out of version control

**Review by inkel — CHANGES_REQUESTED** [review]
Upon first review looks quite good, however, we cannot accept the `replace` directive in `go.mod`.

**Comment by justinTM** [comment]
@inkel i know, it's just to get passing tests until client PR is merged

**Comment by justinTM** [comment]
must be merged together with https://github.com/grafana/grafana-api-golang-client/pull/59

i'll remove the `replace` directive in anticipation

File: grafana/data_source_dashboards.go
Comment (julienduchesne) [inline grafana/data_source_dashboards.go]: I guess we shouldn't have ID conflicts. I think I would hash the query instead. That would allow to have two tag-based datasources for example

File: grafana/data_source_dashboards.go
Comment (julienduchesne) [inline grafana/data_source_dashboards.go]: I'd make this a schema attribute with a 5000 default. Could be useful to find the first matching dashboard for example

**Review by julienduchesne — COMMENTED** [review]
LGTM modulo rebasing on master + the two comments I added

File: grafana/data_source_dashboards.go
Comment (justinTM) [reply grafana/data_source_dashboards.go]: ahhh good idea thanks

**Comment by justinTM** [comment]
https://drone.grafana.net/grafana/terraform-provider-grafana/773

File: grafana/data_source_dashboards.go
Comment (inkel) [inline grafana/data_source_dashboards.go]: Just out of curiosity, any reason why returning just 24 characters?

Comment (inkel) [inline grafana/data_source_dashboards.go]: Also, you could achieve the same by doing:
```suggestion
	return fmt.Sprintf("%x", hashOut.Sum(nil))[:23]
```

**Review by inkel — APPROVED** [review]
LGTM. Thanks for your work and patience!

**Review by julienduchesne — APPROVED** [review]

**Pull request merged**
